### PR TITLE
APT Loss Functions

### PR DESCRIPTION
**Pull request by smericks**
Adding diagonal and full covariance APT loss functions for SNPE training with paltas. Includes test cases.

### CODE REVIEW

File: paltas/Analysis/dataset_generation.py [hunk 1]
```diff
@@ -120,6 +120,37 @@ def unnormalize_outputs(input_norm_path,learning_params,mean,standard_dev=None,
 		if cov_mat is not None:
 			cov_mat[:,lpi,:] *= param_std
 			cov_mat[:,:,lpi] *= param_std
+			
+# TODO: write test after moving (make sure identity operation w/ unnormalized)
+def normalize_mu_prec(mu,prec_mat,input_norm_path):
+    """Helper function to convert mu, prec_matrix to normalized parameter 
+        space
+        
+    Args:
```
Comment (swagnercarena) [inline paltas/Analysis/dataset_generation.py:128]: I'm not familiar with this string formatting. Does that shape information translate to the read the docs?

Remove the indentation from Returns (should have same indentation as Args)

File: paltas/Analysis/dataset_generation.py [hunk 2]
```diff
@@ -120,6 +120,37 @@ def unnormalize_outputs(input_norm_path,learning_params,mean,standard_dev=None,
 		if cov_mat is not None:
 			cov_mat[:,lpi,:] *= param_std
 			cov_mat[:,:,lpi] *= param_std
+			
+# TODO: write test after moving (make sure identity operation w/ unnormalized)
```
Comment (swagnercarena) [inline paltas/Analysis/dataset_generation.py]: I think you've added a test for this now so please remove the TODO.

File: paltas/Analysis/loss_functions.py [hunk 3]
```diff
@@ -425,3 +426,157 @@ def loss(self,y_true,output):
...
+				proposal_prec,input_norm_path)
+
+		# store prior & proposal info which we will need to compute loss
+		self.prior_mu = tf.constant(prior_means,dtype=tf.float32)
+		self.prior_prec = tf.constant(prior_prec,dtype=tf.float32)
+		self.proposal_mu = tf.constant(proposal_means,dtype=tf.float32)
+		self.proposal_prec = tf.constant(proposal_prec,dtype=tf.float32)
+
+	@staticmethod
+	def log_gauss_full(y_true,y_pred,prec_mat):
```
Comment (swagnercarena) [inline paltas/Analysis/loss_functions.py:472]: Is this the same function as the class you're inheriting? If so there's no need to redefine it. Or am I missing an important detail?

File: paltas/Analysis/loss_functions.py [hunk 4]
```diff
@@ -425,3 +426,157 @@ def loss(self,y_true,output):
...
+	 			tf.matmul(self.prior_prec,tf.expand_dims(self.prior_mu,-1)) )
+			mu_comb = tf.matmul(tf.linalg.inv(prec_comb),rhs)
+			# remove extra dimension 
+			mu_comb = tf.squeeze(mu_comb,axis=-1)
+			loss_list.append(self.log_gauss_full(y_true,mu_comb,prec_comb))
+		loss_stack = tf.stack(loss_list,axis=-1)
+		return tf.reduce_min(loss_stack,axis=-1)
+	
+
+class DiagonalCovarianceAPTLoss(DiagonalCovarianceLoss):
```
Comment (swagnercarena) [inline paltas/Analysis/loss_functions.py:519]: I think if you inherit from FullCovarianceAPTLoss you don't have to repeat the __init__

File: test/analysis_tests.py [hunk 5]
```diff
@@ -158,6 +158,32 @@ def test_unormalize_outputs(self):
...
+		# Need to test that the precision matrices output are symmetric!
+		metadata = pd.read_csv(self.fake_test_folder + 'metadata.csv')
+		learning_params = ['main_deflector_parameters_theta_E',
+			'main_deflector_parameters_gamma1','main_deflector_parameters_gamma2',
+			'main_deflector_parameters_gamma']
+		input_norm_path = self.fake_test_folder + 'norms.csv'
+		norm_dict = Analysis.dataset_generation.normalize_outputs(metadata,
+			learning_params,input_norm_path)
+		
+        # Actual mean/precision matrix output by a network for these params
```
Comment (swagnercarena) [inline test/analysis_tests.py:172]: The formatting of the spacing looks odd here

File: test/analysis_tests.py [hunk 6]
```diff
@@ -158,6 +158,32 @@ def test_unormalize_outputs(self):
...
+		proposal_prec = np.asarray([[135114.03  ,  35584.293 ,  -6538.843 ,  -1039.7645],
+            [ 35584.293 ,  43250.586 ,  -1896.0295, -11420.458 ],
+            [ -6538.843 ,  -1896.0295,  62808.117 ,    947.9287],
+            [ -1039.7645, -11420.458 ,    947.9287,   6692.047 ]]).astype(np.float64)
+		
+		mu,prec = Analysis.dataset_generation.normalize_mu_prec(proposal_means,
+			proposal_prec,input_norm_path)
+		
+		# Testing for symmetry of precision matrix
+		self.assertTrue(np.allclose(prec, prec.T))
```
Comment (swagnercarena) [inline test/analysis_tests.py:183]: Is the only check that the precision matrix is symmetric? Could be worth confirming that the normalization was applied correctly.

File: test/analysis_tests.py [hunk 7]
```diff
@@ -1009,6 +1035,414 @@ def test_loss(self):
 		loss = loss_class.loss(yttf,yptf)
 
 		self.assertAlmostEqual(np.sum(loss.numpy()),scipy_nlp,places=4)
+		
+
+class FullCovarianceLossTests(unittest.TestCase):
```
Comment (swagnercarena) [inline test/analysis_tests.py:1040]: Not sure why this is marked as new, but I'm assuming the entire class is the same as before (and therefore doesn't need review).

File: test/analysis_tests.py [hunk 8]
```diff
@@ -1009,6 +1035,414 @@ def test_loss(self):
...
+
+class FullCovarianceAPTLossTests(unittest.TestCase):
+
+	def setUp(self):
+		# Set up a random seed for consistency
+		np.random.seed(2)
+
+	def test_sanity_loss(self):
+
+        # if prior = proposal, check that produces the same as without APT
```
Comment (swagnercarena) [inline test/analysis_tests.py]: Again, something seems odd about the spacing here

File: test/analysis_tests.py [hunk 9]
```diff
@@ -1009,6 +1035,414 @@ def test_loss(self):
...
+			[np.random.normal(size=(batch_size, dim)),
+				np.zeros((batch_size, len_L_mat_elements))], axis=-1
+		)
+
+		truth = np.random.normal(size=(batch_size, dim))
+
+		truth = tf.constant(truth,dtype=tf.float32)
+		outputs = tf.constant(outputs,dtype=tf.float32)
+
+		print(truth.shape)
```
Comment (swagnercarena) [inline test/analysis_tests.py]: Don't want print statements to linger in the tests

File: test/analysis_tests.py [hunk 10]
```diff
@@ -1009,6 +1035,414 @@ def test_loss(self):
...
+         0.46607757,  -0.42691824,  -0.35380697,   1.5319805 ]])
+		truth1 = np.asarray([[ 0.33084044, -0.07505693, -0.7572751 , -0.90663636,  0.34479252,
+       -0.02982552,  1.0541298 , -0.32437885]])
+		output1 = tf.constant(output1,dtype=tf.float32)
+		output1_batched = tf.squeeze(tf.stack([output1,output1]))
+		truth1 = tf.constant(truth1,dtype=tf.float32)
+		truth1_batched = tf.squeeze(tf.stack([truth1,truth1]))
+
+		# prior = proposal, so should be same as gaussian loss
+		loss_diff = np.abs(gaussian_loss1.loss(truth1,output1).numpy()[0] -
```
Comment (swagnercarena) [inline test/analysis_tests.py]: Probably better to do assert_almost_equal and just play with the precision the test requires

File: test/analysis_tests.py [hunk 11]
```diff
@@ -1009,6 +1035,414 @@ def test_loss(self):
...
+			snpe_c_loss1.loss(truth1,output1).numpy()[0])
+		self.assertTrue(loss_diff < 0.001)
+
+		# let's try adding in a batch dimension
+		loss_diff_array = np.abs(gaussian_loss1.loss(truth1_batched,output1_batched).numpy() - 
+			snpe_c_loss1.loss(truth1_batched,output1_batched).numpy())
+		# checking that the difference in loss is never greater than 0.0001
+		self.assertTrue(np.sum(loss_diff_array > 0.001) == 0)
```
Comment (swagnercarena) [inline test/analysis_tests.py]: Similar comment here. Loop over the batch

File: test/analysis_tests.py [hunk 12]
```diff
@@ -1009,6 +1035,414 @@ def test_loss(self):
...
+		# APT prior
+		mu_prior = np.asarray([0.8,0.,0.,2.0,0.,0.,0.,0.])
+		prec_prior = np.linalg.inv(np.diag(np.asarray([0.2,0.12,0.12,0.12,0.18,0.18,0.07,0.07])**2))
+		# APT proposal
+		mu_prop = np.asarray([0.85,-0.01,-0.09,2.07,0.08,-0.02,0.07,-0.02])
+		prec_prop = np.linalg.inv(np.diag(np.asarray([0.001,0.01,0.01,0.11,0.01,0.02,0.003,0.003])**2))
+
+		# create APT loss object
+		input_norm_path = 'test_data/apt_norms.csv'
+		#'/Users/smericks/Desktop/StrongLensing/STRIDES14results/sep7_narrow_lognorm/lr_1e-3/norms.csv'
```
Comment (swagnercarena) [inline test/analysis_tests.py]: Vestigial comment?

File: test/analysis_tests.py [hunk 13]
```diff
@@ -1009,6 +1035,414 @@ def test_loss(self):
...
+		# WARNING: values changed b/c pass by reference
+		snpe_c_loss = Analysis.loss_functions.FullCovarianceAPTLoss(8,
+			mu_prior, prec_prior, mu_prop,prec_prop,input_norm_path=input_norm_path)
+		# move to normalized space
+		# MU_PRIOR, PREC_PRIOR, MU_PROP, PREC_PROP ALREADY MODIFIED B/C PASS BY REFERENCE
+		# (TODO: change how this is handled to avoid further issues)
```
Comment (swagnercarena) [inline test/analysis_tests.py]: I think you fixed this

File: test/analysis_tests.py [hunk 14]
```diff
@@ -1009,6 +1035,414 @@ def test_loss(self):
...
+		# 1.4246841e+03, 6.3971118e+02, 5.4188513e+02 , 2.1357590e+01, 3.5679758e+02,
+		# 4.4313950e+02, 6.8692467e+01, 1.4055161e+02]])
+		truth1 = np.asarray([[ 0.33084044, -0.07505693, -0.7572751 , -0.90663636,  0.34479252,
+       -0.02982552,  1.0541298 , -0.32437885]])
+		output1 = tf.constant(output1,dtype=tf.float32)
+		output1_batched = tf.squeeze(tf.stack([output1,output1]))
+		truth1 = tf.constant(truth1,dtype=tf.float32)
+		truth1_batched = tf.squeeze(tf.stack([truth1,truth1]))
+
+		# CONFIRMED: the problem is NOT the prefactor
```
Comment (swagnercarena) [inline test/analysis_tests.py]: Is this comment still relevant?

File: test/config_tests.py [hunk 15]
```diff
@@ -341,9 +341,11 @@ def draw_subhalos(self,*args,**kwargs):
 		self.assertTrue(np.sum(lens_light_image[90:110,90:110]) >
 			np.sum(image[90:110,90:110]))
 
+		# turn off lens light
+		self.c.lens_light_class = None
 		# Add point source and validate output
-		self.c.sample['point_source_parameters'] = {'x_point_source':0.001,
-			'y_point_source':0.001,'magnitude':22,'output_ab_zeropoint':25.95,
+		self.c.sample['point_source_parameters'] = {'x_point_source':0.0,
```
Comment (swagnercarena) [inline test/config_tests.py]: Why did you change this?

File: test/test_data/apt_norms.csv [hunk 16]
```diff
@@ -0,0 +1,9 @@
+parameter,mean,std
+main_deflector_parameters_theta_E,0.7992458031844867,0.15239538119453927
```
Comment (swagnercarena) [inline test/test_data/apt_norms.csv]: I'm guessing this is pulled from an actual norm file. If you manually change the values to be more exact (0.8, 0.15, etc.) could be easier for testing.

**Review by swagnercarena — COMMENTED** [review]
These changes look excellent, I've just added a few minor comments (mostly about tests).

File: paltas/Analysis/loss_functions.py (hunk 3 above)
Comment (smericks) [reply paltas/Analysis/loss_functions.py:472]: This is not quite the same, the normalization factor is computed differently, because there is no log cholesky decomposition term (L_diag) provided in this case.

File: paltas/Analysis/loss_functions.py (hunk 4 above)
Comment (smericks) [reply paltas/Analysis/loss_functions.py:519]: Yes true, but I need the convert_output() from DiagonalCovarianceLoss, which is not consistent with the convert_output() from FullCovarianceAPT.

File: test/config_tests.py (hunk 15 above)
Comment (smericks) [reply test/config_tests.py]: I switched it back. I'm not sure why but I think this test was failing for me before, but it looks fine to me now.

**Comment by smericks** [comment]
hi @swagnercarena, re-opening this PR for the APT loss functions with long-awaited updates. Chatted with @P-1884 @padmavb-18 and we are all keen on pushing some new updates to paltas since we have some upcoming projects we want to use the code base for
